README.md
# microfft

microfft is a library for computing fast fourier transforms that targets
embedded systems. It provides an in-place implementation of the Radix-2 FFT
algorithm. All computations are performed directly on the input buffer and
require no additional allocations. This makes microfft suitable for `no_std`
environments, like microcontrollers.

Speed is achieved mainly by maintaining a pre-computed sine table that is used
to look up the necessary twiddle factors. By replacing arithmetic operations
with simple memory lookups, we reduce the number of CPU cycles spent.
Unfortunately, the pre-computed table also claims a considerable amount of
memory, which might be a deal-breaker for some embedded projects (see
[Memory Usage](#memory-usage)).

microfft also implements a specialized algorithm for FFTs on real (instead
of complex) values. Naively one would calculate a real FFT simply by converting
the input to complex values (leaving the imaginary part empty) and running a
CFFT. microfft's RFFT algorithm instead packs pairs of real values into
a single complex one each, then computes a CFFT of half the original input
size, followed by some recombination magic. This has the effect of roughly
halving the number of CPU cycles required, as can be seen in the
[benchmark results][bench/README.md].

## Example

The following example demonstrates computing a 16-point RFFT on a set of
samples generated from a sine signal:

```rust
use std::convert::TryInto;
use std::f32::consts::PI;

// generate 16 samples of a sine wave at frequency 3
let sample_count = 16;
let signal_freq = 3.;
let sample_interval = 1. / sample_count as f32;
let mut samples: Vec<_> = (0..sample_count)
    .map(|i| (2. * PI * signal_freq * sample_interval * i as f32).sin())
    .collect();

// compute the RFFT of the samples
let mut samples: [_; 16] = samples.try_into().unwrap();
let mut spectrum = microfft::real::rfft_16(samples);
// since the real-valued coefficient at the Nyquist frequency is packed into the
// imaginary part of the DC bin, it must be cleared before computing the amplitudes
spectrum[0].im = 0.0;

// the spectrum has a spike at index `signal_freq`
let amplitudes: Vec<_> = spectrum.iter().map(|c| c.norm_sqr() as u32).collect();
assert_eq!(&amplitudes, &[0, 0, 0, 64, 0, 0, 0, 0]);
```

## Requirements

Requires Rust version **1.51.0** or newer.

## Sine Tables

microfft keeps a single sine table to calculate the twiddle factors for all
FFT sizes. This removes some memory overhead compared to keeping a separate
table for each FFT size, as there would be duplication between those tables.

The default sine table supports a full 4096-point FFT. If you only want to
compute FFTs of smaller sizes, it is recommended to select the appropriate
`maxn-*` feature, to not waste memory. For example, if your maximum FFT size is
1024, add this to your `Cargo.toml`:

```toml
[dependencies.microfft]
default-features = false
features = ["maxn-1024"]
```

This tells microfft to not provide functions for computing FFTs of sizes larger
than 1024 and to keep only the 1024-point sine table.

## Bit-reversal Tables

The optional feature `bitrev-tables` enables the use of pre-computed tables of
bit-reversed indices required for the reordering of input values performed at
the start of each FFT. If this feature is disabled (the default), the
bit-reversals are computed at runtime instead.

Note that enabling bitrev tables significantly increases the memory usage of
microfft. While it can speed up FFT computation on some systems, there are also
architectures that provide dedicated bit-reversal instructions (like `RBIT` on
ARMv7). On such architectures, switching on bitrev tables is usually
detrimental to performance.

## Limitations

microfft has a few limitations, mostly due to its focus on speed, that might
make it unsuitable for some embedded projects. You should know about these
if you consider using this library:

### Memory Usage

The use of pre-computed sine and bitrev tables means that microfft has
considerable requirements on read-only memory. If your chip doesn't have much
flash to begin with, this can be an issue.

The amount of memory required for tables depends on the the configuration of
the [`maxn-*`](#sine-tables) and [`bitrev-tables`](#bit-reversal-tables)
features:

| `maxn-*`    | without `bitrev-tables` | with `bitrev-tables` |
| ----------- | ----------------------: | -------------------: |
| `maxn-4`    |                       0 |                    8 |
| `maxn-8`    |                       4 |                   20 |
| `maxn-16`   |                      12 |                   44 |
| `maxn-32`   |                      28 |                   92 |
| `maxn-64`   |                      60 |                  188 |
| `maxn-128`  |                     124 |                  380 |
| `maxn-256`  |                     252 |                  764 |
| `maxn-512`  |                     508 |                1,532 |
| `maxn-1024` |                   1,020 |                3,068 |
| `maxn-2048` |                   2,044 |                6,140 |
| `maxn-4096` |                   4,092 |               12,284 |

In addition, the code size also increases with FFT size.

### Supported FFT Sizes

microfft only supports FFT point-sizes that are powers of two, a limitation of
the Radix-2 algorithm. Additionally, the maximum supported size is currently
4096, although this limit can be increased in the future as necessary.

### `f64` Support

This library currently only supports single-precision floating-point inputs.
Similarly to the FFT size limit, this is a restriction that might be lifted
in the future, should the need arise.

## License

This project is licensed under the MIT license ([LICENSE](LICENSE) or
http://opensource.org/licenses/MIT).

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in microfft by you, shall be licensed as above, without any
additional terms or conditions.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 0.5.1 (unreleased)

### Added

- Support for inverse FFTs.


## 0.5.0 (2022-06-19)

### Changed

- **Breaking:** The MSRV has been increased to 1.56.0.
- **Breaking:** The former `maxn-*` features of this crate have been renamed
  to `size-*`. They still work basically the same way but are now additive,
  which means it is allowed that more than one of them is enabled at the same
  time. This makes microfft work as expected in the face of Cargo's feature
  unification.

### Added

- A `std` feature that is disabled by default. It transitively enables the
  `std` feature on the `num-complex` dependency, enabling more methods on the
  `Complex32` values returned by the FFT functions.
- Support for FFT sizes 8192 and 16384.


## 0.4.0 (2021-04-03)

### Changed

- **Breaking:** The MSRV has been increased to 1.51.0.
- **Breaking:** All API functions for computing FFTs now take references to
  arrays instead of references to slices. This has the benefit of moving the
  length checking to compile time, which makes those functions panic-proof.
- **Breaking:** When computing RFFTs, the real-valued coefficient at the
  Nyquist frequency is now packed into the imaginary part of the DC bin
  (bin 0). Previously this value was simply dropped from the RFFT output.
- Added `#[must_use]` annotations to all FFT API functions.

### Fixed

- Thanks to Cargo's new resolver it is now possible to build microfft as a
  stand-alone library.


## 0.3.1 (2020-11-09)

### Fixed

- Fixed a bug during the RFFT recombination calculation that caused a wrong
  output value in bin `N / 4`.


## 0.3.0 (2020-03-08)

### Changed

- Store only the largest sine table, instead of one for each FFT size. By
  default this is the table for the 4096-point FFT (the largest supported one),
  but smaller ones can be selected via the new `maxn-*` crate features.


## 0.2.0 (2020-03-08)

### Changed

- Bitrev tables are not used anymore by default, instead the bit-reversed
  indices are computed directly at runtime. This significantly reduces the
  memory usage of microfft. On architectures that provide a dedicated
  bit-reversal instruction (like `RBIT` on ARMv7), speed is also increased.
  The `bitrev-tables` feature can be enabled to still opt into using bitrev
  tables.


## 0.1.2 (2020-03-07)

### Changed

- Store pre-computed sine values instead of full twiddles, reducing the size
  of the twiddle tables to one fourth the prior size.

## 0.1.1 (2020-03-05)

### Added

- Support for FFT sizes 2048 and 4096.


## 0.1.0 (2020-03-04)

### Added

- Support for complex and real FFTs up to size 1024.


tests/transform.rs
use std::convert::TryInto;

use microfft::Complex32;
use rustfft::{algorithm::Radix4, Fft, FftDirection};

fn rust_fft(input: &[Complex32]) -> Vec<Complex32> {
    // Convert to rustfft's `num_complex` types, to prevent issues with
    // incompatible versions.
    let mut buf: Vec<_> = input
        .iter()
        .map(|c| rustfft::num_complex::Complex32::new(c.re, c.im))
        .collect();

    let fft = Radix4::new(buf.len(), FftDirection::Forward);
    fft.process(&mut buf);

    buf.iter().map(|c| Complex32::new(c.re, c.im)).collect()
}

fn approx_eq(a: Complex32, b: Complex32) -> bool {
    fn approx_f32(x: f32, y: f32) -> bool {
        let diff = (x - y).abs();
        let rel_diff = if x != 0. { (diff / x).abs() } else { diff };
        rel_diff < 0.02
    }

    approx_f32(a.re, b.re) && approx_f32(a.im, b.im)
}

fn assert_approx_eq(xa: &[Complex32], xb: &[Complex32]) {
    assert_eq!(xa.len(), xb.len());
    for (a, b) in xa.iter().zip(xb) {
        assert!(approx_eq(*a, *b));
    }
}

macro_rules! cfft_tests {
    ( $( $name:ident: $N:expr, )* ) => {
        $(
            #[test]
            fn $name() {
                let input: Vec<_> = (0..$N)
                    .map(|i| i as f32)
                    .map(|f| Complex32::new(f, f))
                    .collect();

                let expected = rust_fft(&input);
                let input: [_; $N] = input.try_into().unwrap();
                let result = microfft::complex::$name(input);

                assert_approx_eq(&result, &expected);
            }
        )*
    };
}

cfft_tests! {
    cfft_2: 2,
    cfft_4: 4,
    cfft_8: 8,
    cfft_16: 16,
    cfft_32: 32,
    cfft_64: 64,
    cfft_128: 128,
    cfft_256: 256,
    cfft_512: 512,
    cfft_1024: 1024,
    cfft_2048: 2048,
    cfft_4096: 4096,
}

macro_rules! rfft_tests {
    ( $( $name:ident: ($N:expr, $cfft_name:ident), )* ) => {
        $(
            #[test]
            fn $name() {
                let input: Vec<_> = (5..($N+5)).map(|i| i as f32).collect();
                let input_c: Vec<_> = input.iter().map(|f| Complex32::new(*f, 0.)).collect();

                let input_c: [_; $N] = input_c.try_into().unwrap();
                let expected = microfft::complex::$cfft_name(input_c);
                let input: [_; $N] = input.try_into().unwrap();
                let mut result = microfft::real::$name(input);
                // The real-valued coefficient at the Nyquist frequency
                // is packed into the imaginary part of the DC bin.
                let coeff_at_nyquist = result[0].im;
                assert_eq!(coeff_at_nyquist, expected[$N / 2].re);
                // Clear this value before checking the results.
                result[0].im = 0.0;
                assert_approx_eq(&result, &expected[..($N / 2)]);
            }
        )*
    };
}

rfft_tests! {
    rfft_2: (2, cfft_2),
    rfft_4: (4, cfft_4),
    rfft_8: (8, cfft_8),
    rfft_16: (16, cfft_16),
    rfft_32: (32, cfft_32),
    rfft_64: (64, cfft_64),
    rfft_128: (128, cfft_128),
    rfft_256: (256, cfft_256),
    rfft_512: (512, cfft_512),
    rfft_1024: (1024, cfft_1024),
    rfft_2048: (2048, cfft_2048),
    rfft_4096: (4096, cfft_4096),
}


src/cfft.rs
use crate::{tables, Complex32};

pub(crate) trait CFft {
    type Half: CFft;

    const N: usize;
    const LOG2_N: usize;

    #[cfg(feature = "bitrev-tables")]
    const BITREV_TABLE: &'static [u16] = tables::BITREV[Self::LOG2_N];

    #[inline]
    fn transform_slice(x: &mut [Complex32]) {
        debug_assert_eq!(x.len(), Self::N);

        Self::bit_reverse_reorder(x);
        Self::compute_butterflies(x);
    }

    #[cfg(feature = "bitrev-tables")]
    #[inline]
    fn bit_reverse_reorder(x: &mut [Complex32]) {
        debug_assert_eq!(x.len(), Self::N);

        for i in 0..Self::N {
            let j = Self::BITREV_TABLE[i] as usize;
            x.swap(i, j);
        }
    }

    #[cfg(not(feature = "bitrev-tables"))]
    #[inline]
    fn bit_reverse_reorder(x: &mut [Complex32]) {
        debug_assert_eq!(x.len(), Self::N);

        let shift = core::mem::size_of::<usize>() * 8 - Self::LOG2_N;
        for i in 0..Self::N {
            let rev = i.reverse_bits();
            let j = rev >> shift;
            if j > i {
                x.swap(i, j);
            }
        }
    }

    #[inline]
    fn compute_butterflies(x: &mut [Complex32]) {
        debug_assert_eq!(x.len(), Self::N);

        let m = Self::N / 2;
        let u = m / 2;

        let table_len = tables::SINE.len();
        let table_stride = (table_len + 1) * 4 / Self::N;

        Self::Half::compute_butterflies(&mut x[..m]);
        Self::Half::compute_butterflies(&mut x[m..]);

        // [k = 0] twiddle factor: `1 + 0i`
        let (x_0, x_m) = (x[0], x[m]);
        x[0] = x_0 + x_m;
        x[m] = x_0 - x_m;

        // [k in [1, m/2)] twiddle factor:
        //   - re from SINE table backwards and negative
        //   - im from SINE table directly
        for k in 1..u {
            let s = k * table_stride;
            let re = tables::SINE[table_len - s] * -1.;
            let im = tables::SINE[s - 1];
            let twiddle = Complex32::new(re, im);

            let (x_k, x_km) = (x[k], x[k + m]);
            let y = twiddle * x_km;
            x[k] = x_k + y;
            x[k + m] = x_k - y;
        }

        // [k = m/2] twiddle factor: `0 - 1i`
        let (x_u, x_um) = (x[u], x[u + m]);
        let y = x_um * Complex32::new(0., -1.);
        x[u] = x_u + y;
        x[u + m] = x_u - y;

        // [k in (m/2, m)] twiddle factor:
        //   - re from SINE table directly
        //   - im from SINE table backwards
        for k in (u + 1)..m {
            let s = (k - u) * table_stride;
            let re = tables::SINE[s - 1];
            let im = tables::SINE[table_len - s];
            let twiddle = Complex32::new(re, im);

            let (x_k, x_km) = (x[k], x[k + m]);
            let y = twiddle * x_km;
            x[k] = x_k + y;
            x[k + m] = x_k - y;
        }
    }
}

pub(crate) struct CFftN1;

impl CFft for CFftN1 {
    type Half = Self;

    const N: usize = 1;
    const LOG2_N: usize = 0;

    #[inline]
    fn bit_reverse_reorder(x: &mut [Complex32]) {
        debug_assert_eq!(x.len(), 1);
    }

    #[inline]
    fn compute_butterflies(x: &mut [Complex32]) {
        debug_assert_eq!(x.len(), 1);
    }
}

pub(crate) struct CFftN2;

impl CFft for CFftN2 {
    type Half = CFftN1;

    const N: usize = 2;
    const LOG2_N: usize = 1;

    #[inline]
    fn compute_butterflies(x: &mut [Complex32]) {
        debug_assert_eq!(x.len(), 2);

        let (x_0, x_1) = (x[0], x[1]);
        x[0] = x_0 + x_1;
        x[1] = x_0 - x_1;
    }
}

impl CFftN2 {
    #[inline]
    pub(crate) fn transform(mut x: [Complex32; 2]) -> [Complex32; 2] {
        Self::transform_slice(&mut x);
        x
    }
}

macro_rules! cfft_impls {
    ( $( $I:expr => ($N:expr, $CFftN:ident, $Half:ident), )* ) => {
        $(
            #[allow(dead_code)]
            pub(crate) struct $CFftN;

            impl CFft for $CFftN {
                type Half = $Half;

                const N: usize = $N;
                const LOG2_N: usize = $I;
            }

            impl $CFftN {
                #[inline]
                pub(crate) fn transform(mut x: [Complex32; $N]) -> [Complex32; $N] {
                    Self::transform_slice(&mut x);
                    x
                }
            }
        )*
    };
}

cfft_impls! {
     2 => (4, CFftN4, CFftN2),
     3 => (8, CFftN8, CFftN4),
     4 => (16, CFftN16, CFftN8),
     5 => (32, CFftN32, CFftN16),
     6 => (64, CFftN64, CFftN32),
     7 => (128, CFftN128, CFftN64),
     8 => (256, CFftN256, CFftN128),
     9 => (512, CFftN512, CFftN256),
    10 => (1024, CFftN1024, CFftN512),
    11 => (2048, CFftN2048, CFftN1024),
    12 => (4096, CFftN4096, CFftN2048),
}


src/real.rs
//! FFT on real inputs (RFFT)
//!
//! This implementation takes advantage of the symmetry properties of
//! the FFT to compute an `N`-point RFFT by internally invoking an
//! `N/2`-point CFFT, roughly doubling the computation speed compared
//! to used a full `N`-point CFFT.
//!
//! The produced output is the first half out the output returned by
//! the corresponding `N`-point CFFT, i.e. the real DC value and
//! `N/2 - 1` positive-frequency terms. Additionally, the real-valued
//! coefficient at the Nyquist frequency is packed into the imaginary
//! part of the DC bin. The negative-frequency terms are not computed,
//! since they can be calculated from the positive-frequency terms and
//! are therefore redundant.

use crate::{rfft::*, Complex32};

/// Perform an in-place 2-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_2;
///
/// let input = [0.; 2];
/// let result = rfft_2(input);
/// ```
#[inline]
#[must_use]
pub fn rfft_2(input: [f32; 2]) -> [Complex32; 1] {
    RFftN2::transform(input)
}

/// Perform an in-place 4-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_4;
///
/// let input = [0.; 4];
/// let result = rfft_4(input);
/// ```
#[inline]
#[must_use]
pub fn rfft_4(input: [f32; 4]) -> [Complex32; 2] {
    RFftN4::transform(input)
}

/// Perform an in-place 8-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_8;
///
/// let input = [0.; 8];
/// let result = rfft_8(input);
/// ```
#[cfg(any(
    feature = "maxn-8",
    feature = "maxn-16",
    feature = "maxn-32",
    feature = "maxn-64",
    feature = "maxn-128",
    feature = "maxn-256",
    feature = "maxn-512",
    feature = "maxn-1024",
    feature = "maxn-2048",
    feature = "maxn-4096",
))]
#[inline]
#[must_use]
pub fn rfft_8(input: [f32; 8]) -> [Complex32; 4] {
    RFftN8::transform(input)
}

/// Perform an in-place 16-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_16;
///
/// let input = [0.; 16];
/// let result = rfft_16(input);
/// ```
#[cfg(any(
    feature = "maxn-16",
    feature = "maxn-32",
    feature = "maxn-64",
    feature = "maxn-128",
    feature = "maxn-256",
    feature = "maxn-512",
    feature = "maxn-1024",
    feature = "maxn-2048",
    feature = "maxn-4096",
))]
#[inline]
#[must_use]
pub fn rfft_16(input: [f32; 16]) -> [Complex32; 8] {
    RFftN16::transform(input)
}

/// Perform an in-place 32-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_32;
///
/// let input = [0.; 32];
/// let result = rfft_32(input);
/// ```
#[cfg(any(
    feature = "maxn-32",
    feature = "maxn-64",
    feature = "maxn-128",
    feature = "maxn-256",
    feature = "maxn-512",
    feature = "maxn-1024",
    feature = "maxn-2048",
    feature = "maxn-4096",
))]
#[inline]
#[must_use]
pub fn rfft_32(input: [f32; 32]) -> [Complex32; 16] {
    RFftN32::transform(input)
}

/// Perform an in-place 64-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_64;
///
/// let input = [0.; 64];
/// let result = rfft_64(input);
/// ```
#[cfg(any(
    feature = "maxn-64",
    feature = "maxn-128",
    feature = "maxn-256",
    feature = "maxn-512",
    feature = "maxn-1024",
    feature = "maxn-2048",
    feature = "maxn-4096",
))]
#[inline]
#[must_use]
pub fn rfft_64(input: [f32; 64]) -> [Complex32; 32] {
    RFftN64::transform(input)
}

/// Perform an in-place 128-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_128;
///
/// let input = [0.; 128];
/// let result = rfft_128(input);
/// ```
#[cfg(any(
    feature = "maxn-128",
    feature = "maxn-256",
    feature = "maxn-512",
    feature = "maxn-1024",
    feature = "maxn-2048",
    feature = "maxn-4096",
))]
#[inline]
#[must_use]
pub fn rfft_128(input: [f32; 128]) -> [Complex32; 64] {
    RFftN128::transform(input)
}

/// Perform an in-place 256-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_256;
///
/// let input = [0.; 256];
/// let result = rfft_256(input);
/// ```
#[cfg(any(
    feature = "maxn-256",
    feature = "maxn-512",
    feature = "maxn-1024",
    feature = "maxn-2048",
    feature = "maxn-4096",
))]
#[inline]
#[must_use]
pub fn rfft_256(input: [f32; 256]) -> [Complex32; 128] {
    RFftN256::transform(input)
}

/// Perform an in-place 512-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_512;
///
/// let input = [0.; 512];
/// let result = rfft_512(input);
/// ```
#[cfg(any(
    feature = "maxn-512",
    feature = "maxn-1024",
    feature = "maxn-2048",
    feature = "maxn-4096",
))]
#[inline]
#[must_use]
pub fn rfft_512(input: [f32; 512]) -> [Complex32; 256] {
    RFftN512::transform(input)
}

/// Perform an in-place 1024-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_1024;
///
/// let input = [0.; 1024];
/// let result = rfft_1024(input);
/// ```
#[cfg(any(feature = "maxn-1024", feature = "maxn-2048", feature = "maxn-4096"))]
#[inline]
#[must_use]
pub fn rfft_1024(input: [f32; 1024]) -> [Complex32; 512] {
    RFftN1024::transform(input)
}

/// Perform an in-place 2048-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_2048;
///
/// let input = [0.; 2048];
/// let result = rfft_2048(input);
/// ```
#[cfg(any(feature = "maxn-2048", feature = "maxn-4096"))]
#[inline]
#[must_use]
pub fn rfft_2048(input: [f32; 2048]) -> [Complex32; 1024] {
    RFftN2048::transform(input)
}

/// Perform an in-place 4096-point RFFT.
///
/// # Example
///
/// ```
/// use microfft::real::rfft_4096;
///
/// let input = [0.; 4096];
/// let result = rfft_4096(input);
/// ```
#[cfg(any(feature = "maxn-4096"))]
#[inline]
#[must_use]
pub fn rfft_4096(input: [f32; 4096]) -> [Complex32; 2048] {
    RFftN4096::transform(input)
}


src/rfft.rs
use crate::{cfft::*, tables, Complex32};
use core::mem;
use static_assertions::{assert_eq_align, assert_eq_size};

pub(crate) trait RFft {
    type CFft: CFft;

    const N: usize = Self::CFft::N * 2;

    #[inline]
    fn transform_slice(x: &mut [Complex32]) {
        debug_assert_eq!(x.len(), Self::CFft::N);

        Self::CFft::transform_slice(x);
        Self::recombine(x);
    }

    #[inline]
    fn recombine(x: &mut [Complex32]) {
        let m = Self::CFft::N;
        debug_assert_eq!(x.len(), m);

        let table_len = tables::SINE.len();
        let table_stride = (table_len + 1) * 4 / Self::N;

        // The real part of the first element is the DC value.
        // Additionally, the real-valued coefficient at the Nyquist frequency
        // is stored in the imaginary part.
        let x0 = x[0];
        x[0] = Complex32::new(x0.re + x0.im, x0.re - x0.im);

        let u = m / 2;
        for k in 1..u {
            let s = k * table_stride;
            let twiddle_re = tables::SINE[table_len - s] * -1.;
            let twiddle_im = tables::SINE[s - 1];

            let (x_k, x_nk) = (x[k], x[m - k]);
            let sum = (x_k + x_nk) / 2.;
            let diff = (x_k - x_nk) / 2.;

            x[k] = Complex32::new(
                sum.re + twiddle_re * sum.im + twiddle_im * diff.re,
                diff.im + twiddle_im * sum.im - twiddle_re * diff.re,
            );
            x[m - k] = Complex32::new(
                sum.re - twiddle_re * sum.im - twiddle_im * diff.re,
                -diff.im + twiddle_im * sum.im - twiddle_re * diff.re,
            );
        }

        let xu = x[u];
        x[u] = Complex32::new(xu.re, -xu.im);
    }
}


pub(crate) struct RFftN2;

impl RFft for RFftN2 {
    type CFft = CFftN1;

    #[inline]
    fn recombine(x: &mut [Complex32]) {
        debug_assert_eq!(x.len(), 1);

        // The real part of the first element is the DC value.
        // Additionally, the real-valued coefficient at the Nyquist frequency
        // is stored in the imaginary part.
        let x0 = x[0];
        x[0] = Complex32::new(x0.re + x0.im, x0.re - x0.im);
    }
}

impl RFftN2 {
    #[inline]
    pub(crate) fn transform(x: [f32; 2]) -> [Complex32; 1] {
        let mut x = Self::pack_complex(x);
        Self::transform_slice(&mut x);
        x
    }

    #[inline]
    fn pack_complex(x: [f32; 2]) -> [Complex32; 1] {
        assert_eq_size!([f32; 2], [Complex32; 1]);
        assert_eq_align!([f32; 2], [Complex32; 1]);
        unsafe { mem::transmute(x) }
    }
}

macro_rules! rfft_impls {
    ( $( ($RFftN:ident, $CFftN:ident), )* ) => {
        $(
            #[allow(dead_code)]
            pub(crate) struct $RFftN;

            impl RFft for $RFftN {
                type CFft = $CFftN;
            }

            impl $RFftN {
                #[inline]
                pub(crate) fn transform(x: [f32; $RFftN::N]) -> [Complex32; $CFftN::N] {
                    let mut x = Self::pack_complex(x);
                    Self::transform_slice(&mut x);
                    x
                }

                #[inline]
                fn pack_complex(x: [f32; $RFftN::N]) -> [Complex32; $CFftN::N] {
                    assert_eq_size!([f32; $RFftN::N], [Complex32; $CFftN::N]);
                    assert_eq_align!([f32; $RFftN::N], [Complex32; $CFftN::N]);
                    unsafe { mem::transmute(x) }
                }
            }
        )*
    };
}

rfft_impls! {
    (RFftN4, CFftN2),
    (RFftN8, CFftN4),
    (RFftN16, CFftN8),
    (RFftN32, CFftN16),
    (RFftN64, CFftN32),
    (RFftN128, CFftN64),
    (RFftN256, CFftN128),
    (RFftN512, CFftN256),
    (RFftN1024, CFftN512),
    (RFftN2048, CFftN1024),
    (RFftN4096, CFftN2048),
}


bench/src/main.rs
#![no_std]
#![no_main]
#![allow(dead_code)]

use cortex_m::iprint;
use cortex_m_rt::entry;
use hal::{prelude::*, time::MonoTimer};
use microfft::Complex32;
use panic_semihosting as _;

#[entry]
fn main() -> ! {
    let core = cortex_m::Peripherals::take().unwrap();
    let device = hal::stm32::Peripherals::take().unwrap();

    let mut flash = device.FLASH.constrain();
    let rcc = device.RCC.constrain();
    let clocks = rcc
        .cfgr
        .use_hse(8.mhz())
        .sysclk(72.mhz())
        .hclk(72.mhz())
        .pclk1(36.mhz())
        .pclk2(72.mhz())
        .freeze(&mut flash.acr);

    let timer = MonoTimer::new(core.DWT, clocks);
    let cycles = bench::run(timer);

    let mut itm = core.ITM;
    iprint!(&mut itm.stim[0], "{}", cycles);

    panic!("bench done");
}

#[cfg(feature = "microfft-c")]
mod bench {
    use super::{n, timeit, Complex32, MonoTimer};

    pub fn run(timer: MonoTimer) -> u32 {
        let mut x = [Complex32::default(); n::N];
        for (i, c) in x.iter_mut().enumerate() {
            *c = Complex32::new(i as f32, 0.);
        }

        timeit(timer, || n::CFFT(x))
    }
}

#[cfg(feature = "microfft-r")]
mod bench {
    use super::{n, timeit, MonoTimer};

    pub fn run(timer: MonoTimer) -> u32 {
        let mut x = [0.; n::N];
        for (i, c) in x.iter_mut().enumerate() {
            *c = i as f32;
        }

        timeit(timer, || n::RFFT(x))
    }
}

fn timeit<F, R>(timer: MonoTimer, f: F) -> u32
where
    F: FnOnce() -> R,
{
    let instant = timer.now();
    let _ = f();
    instant.elapsed()
}

type FnCfft<const N: usize> = fn([Complex32; N]) -> [Complex32; N];
type FnRfft<const N: usize, const M: usize> = fn([f32; N]) -> [Complex32; M];

#[cfg(feature = "n-4")]
mod n {
    pub const N: usize = 4;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_4;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_4;
}

#[cfg(feature = "n-8")]
mod n {
    pub const N: usize = 8;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_8;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_8;
}

#[cfg(feature = "n-16")]
mod n {
    pub const N: usize = 16;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_16;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_16;
}

#[cfg(feature = "n-32")]
mod n {
    pub const N: usize = 32;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_32;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_32;
}

#[cfg(feature = "n-64")]
mod n {
    pub const N: usize = 64;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_64;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_64;
}

#[cfg(feature = "n-128")]
mod n {
    pub const N: usize = 128;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_128;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_128;
}

#[cfg(feature = "n-256")]
mod n {
    pub const N: usize = 256;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_256;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_256;
}

#[cfg(feature = "n-512")]
mod n {
    pub const N: usize = 512;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_512;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_512;
}

#[cfg(feature = "n-1024")]
mod n {
    pub const N: usize = 1024;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_1024;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_1024;
}

#[cfg(feature = "n-2048")]
mod n {
    pub const N: usize = 2048;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_2048;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_2048;
}

#[cfg(feature = "n-4096")]
mod n {
    pub const N: usize = 4096;
    pub const CFFT: super::FnCfft<N> = microfft::complex::cfft_4096;
    pub const RFFT: super::FnRfft<N, {N / 2}> = microfft::real::rfft_4096;
}


src/lib.rs
//! Library for computing fast fourier transforms on embedded systems.
//!
//! microfft provides an in-place implementation of the Radix-2 FFT algorithm.
//! All computations are performed directly on the input buffer and require no
//! additional allocations. This makes microfft suitable for `no_std`
//! environments.
//!
//! In addition to the standard FFT implementation on [`Complex32`] values
//! ([`complex`]), an implementation working on real (`f32`) input values is
//! provided ([`real`]). An `N`-point RFFT internally computes an `N/2`-point
//! CFFT, making it roughly twice as fast as the complex variant.
//!
//! # Example
//!
//! ```
//! use std::convert::TryInto;
//! use std::f32::consts::PI;
//!
//! // generate 16 samples of a sine wave at frequency 3
//! let sample_count = 16;
//! let signal_freq = 3.;
//! let sample_interval = 1. / sample_count as f32;
//! let mut samples: Vec<_> = (0..sample_count)
//!     .map(|i| (2. * PI * signal_freq * sample_interval * i as f32).sin())
//!     .collect();
//!
//! // compute the RFFT of the samples
//! let mut samples: [_; 16] = samples.try_into().unwrap();
//! let mut spectrum = microfft::real::rfft_16(samples);
//! // since the real-valued coefficient at the Nyquist frequency is packed into the
//! // imaginary part of the DC bin, it must be cleared before computing the amplitudes
//! spectrum[0].im = 0.0;
//!
//! // the spectrum has a spike at index `signal_freq`
//! let amplitudes: Vec<_> = spectrum.iter().map(|c| c.norm_sqr() as u32).collect();
//! assert_eq!(&amplitudes, &[0, 0, 0, 64, 0, 0, 0, 0]);
//! ```
//!
//! [`complex`]: complex/index.html
//! [`real`]: real/index.html
//! [`Complex32`]: type.Complex32.html

#![no_std]
#![deny(missing_docs)]
#![warn(rust_2018_idioms)]

pub mod complex;
pub mod real;

pub use num_complex::Complex32;

mod cfft;
mod rfft;
mod tables;

use static_assertions::assert_cfg;

assert_cfg!(
    any(
        feature = "maxn-4",
        feature = "maxn-8",
        feature = "maxn-16",
        feature = "maxn-32",
        feature = "maxn-64",
        feature = "maxn-128",
        feature = "maxn-256",
        feature = "maxn-512",
        feature = "maxn-1024",
        feature = "maxn-2048",
        feature = "maxn-4096",
    ),
    "One of the `maxn-*` features of this crate must be set."
);
